File: mbw_dms/mbw_dms/report/báo_cáo_công_nợ_theo_đơn_hàng/báo_cáo_công_nợ_theo_đơn_hàng.py
import frappe
from frappe import _

def execute(filters=None):
    # Khởi tạo tiêu đề các cột trong báo cáo
    columns = [
        {
            "label": _("Chọn"),
            "fieldname": "select_item",
            "fieldtype": "Check",
            "width": 70
        },
        {
            "label": _("Ngày giao DN"),
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 150
        },
    	{
            "label": _("Tên khách hàng"),
            "fieldname": "customer_name",
            "fieldtype": "Link",
            "options": "Customer",
            "width": 200
        },
        {
            "label": _("Mã SO"),
            "fieldname": "against_sales_order",
            "fieldtype": "Link",
            "options": "Sales Order",
            "width": 220
        },
        {
            "label": "Ngày đặt hàng",
            "fieldname": "transaction_date",
            "fieldtype": "Date",
            "width": 200,
            "name": "Ngày đặt hàng",
        },
        {
            "label": _("Số tiền"),
            "fieldname": "total",
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "label": _("Thanh toán"),
            "fieldname": "trangthaithanhtoan",
            "fieldtype": "Data",
            "width": 200
        },
        {
            "label": _("Mã đơn DN"),
            "fieldname": "name",
            "fieldtype": "Link",
            "options": "Delivery Note",
            "width": 200
        }
    ]

    if filters.get("thanhtoan") == 1:
        filters["thanhtoan"] = 0
    
    # Xây dựng điều kiện tìm kiếm
    conditions = ""
    if filters.get("customer"):
        customer_filter = filters.get("customer")
        if isinstance(customer_filter, list):
            customer_placeholders = ', '.join(['%s'] * len(customer_filter))
            conditions += f" AND dn.customer_name IN ({customer_placeholders})"
        else:
            conditions += " AND dn.customer_name LIKE %s"
    if filters.get("selected_note") and filters.get("delivery_note"):
        deliverynote_filter = filters.get("delivery_note")
        if isinstance(deliverynote_filter, list):
            deliverynote = ', '.join(['%s'] * len(deliverynote_filter))
            conditions += f" AND dn.name IN ({deliverynote})"
    if filters.get("deliverynote"):
        conditions += " AND dn.name LIKE %s"
    if filters.get("thanhtoan") is not None:
        conditions += " AND dn.da_thanh_toan = %s"

    # Truy vấn dữ liệu từ bảng Sales Order và Delivery Note
    query = """
	SELECT 
	    dn.customer_name,
	    dni.against_sales_order,
	    dn.total,
	    CASE 
		WHEN dn.da_thanh_toan = 1 THEN 'Đã thanh toán' 
		ELSE 'Chưa thanh toán' 
	    END AS "trangthaithanhtoan", 
	    dn.name,
	    dn.posting_date
	FROM
	    `tabDelivery Note` dn
	JOIN
	    `tabDelivery Note Item` dni ON dn.name = dni.parent
        WHERE
            dn.docstatus = 1
            AND dn.posting_date BETWEEN %s AND %s
            {conditions}
        GROUP BY
	    dn.name
	ORDER BY
	    dn.posting_date DESC;
    """.format(conditions=conditions)

    # Xây dựng danh sách tham số
    params = [filters.get("from_date"), filters.get("to_date")]
    if filters.get("customer"):
        params.append("%" + filters.get("customer") + "%")
    if filters.get("deliverynote"):
        params.append("%" + filters.get("deliverynote") + "%")
    if filters.get("thanhtoan") is not None:
        params.append(int(filters.get("thanhtoan")))
    if filters.get("selected_note") and filters.get("delivery_note"):
        deliverynote_filter = filters.get("delivery_note")
        if isinstance(deliverynote_filter, list):
            params.extend(deliverynote_filter)
    data = frappe.db.sql(query, params, as_dict=True)
    
    # Bổ sung cột checkbox cho từng dòng trong dữ liệu
    for row in data:
        row['select_item'] = 0 # Checkbox không bị chọn mặc định và có thể chọn được
        sales_order = frappe.db.get_value("Sales Order", row['against_sales_order'],['transaction_date'])
        row['transaction_date'] = sales_order
    total_amount = sum([row['total'] for row in data])
    return columns, data

File: mbw_dms/mbw_dms/report/báo_cáo_công_nợ_theo_đơn_hàng/báo_cáo_công_nợ_theo_đơn_hàng.js
let now = new Date();
let checkedCheckboxes = {};
let Total = 0
let arraydDeliveryNote = [];
// Custom remove function to remove elements from the array
Array.prototype.remove = function(value) {
    let index = this.indexOf(value);
    if (index !== -1) {
        this.splice(index, 1);
    }
};
frappe.query_reports["Báo cáo công nợ theo đơn hàng"] = {
    "filters": [
        {
            fieldname: "from_date",
            label: __("Từ ngày"),
            fieldtype: "Date",
            width: "80",
            reqd: 1,
            default: new Date(now.getFullYear(), now.getMonth(), 1),
        },
        {
            fieldname: "to_date",
            label: __("Đến ngày"),
            fieldtype: "Date",
            width: "80",
            reqd: 1,
            default: new Date(now.getFullYear(), now.getMonth() + 1, 0),
        },
        {
            fieldname: "customer",
            label: __("Khách hàng"),
            fieldtype: "Link",
            width: "80",
            options: "Customer"
        },
        {
            fieldname: "deliverynote",
            label: __("Mã đơn DN"),
            fieldtype: "Link",
            width: "80",
            options: "Delivery Note"
        },
        {
            fieldname: "thanhtoan",
            label: __("Chưa Thanh toán"),
            fieldtype: "Check"
        }
    ],

    // This runs when the report page is loaded
    "onload": function(report) {
        // Add a Print button to the report page
        report.page.add_button(__('In dữ liệu'), function() {
            // Use frappe.ui.get_print_settings with a callback for printing
            frappe.ui.get_print_settings(false, (print_settings) => {
                // Call the print function after getting the settings
                print_report(report, print_settings);
            });
        }, 'Print');
    },

    "after_datatable_render": function(report) {
        $('.dt-row-filter').remove();
        checkedCheckboxes={};
        Total = 0
        arraydDeliveryNote = [];
        // Function to enable checkboxes
        function enableCheckboxes() {
            $(".datatable").find('input[type="checkbox"]').each(function() {
                $(this).val($(this).closest('.dt-cell').attr("data-row-index"));

                $(this).prop('disabled', false);  // Enable checkbox
                $(this).removeClass('disabled-deselected');  // Remove any class indicating disabled state
                
                // Set checkbox state based on previously stored values
                if (checkedCheckboxes[$(this).val()]) {
                    $(this).prop('checked', true);
                }

                $(this).off('click').on('click', function() {
                    // Handle the click event
                    if ($(this).is(':checked')) {
                        checkedCheckboxes[$(this).val()] = true;
                        arraydDeliveryNote.push($(this).closest('.dt-row').find("a[data-doctype='Delivery Note']").html());
                        Total += report.datamanager.data[$(this).val()].total

                    } else {
                        delete checkedCheckboxes[$(this).val()];
                        arraydDeliveryNote.remove($(this).closest('.dt-row').find("a[data-doctype='Delivery Note']").html());
                        Total -= report.datamanager.data[$(this).val()].total
                    }
                });
            });

        }

        // Enable checkboxes initially after rendering the datatable
        enableCheckboxes();

        // Add a scroll event listener to enable checkboxes while scrolling
        $(".datatable").find('.dt-scrollable').on('scroll', function() {
            enableCheckboxes();
        });
    }
};

// Print report function
function print_report(report, print_settings) {
    let columns = report.get_visible_columns ? report.get_visible_columns() : [];
    columns = columns.filter(column => column.fieldname !== 'name' && column.fieldname !== 'posting_date');
    let columns_date = {
        "label": "Ngày đặt hàng",
        "fieldname": "transaction_date",
        "fieldtype": "Date",
        "width": 200,
        "id": "transaction_date",
        "name": "Ngày đặt hàng",
        "editable": false,
        }
    columns.splice(3, 0, columns_date);
    let filtered_data = report.data.filter(row => {
        return arraydDeliveryNote.includes(row.name);
    }).map(({ name, posting_date, trangthaithanhtoan, ...rest })  => {
        if (trangthaithanhtoan === "Chưa thanh toán") {
            trangthaithanhtoan = "CTT";
        }
        if (trangthaithanhtoan === "Đã thanh toán") {
            trangthaithanhtoan = "ĐTT";
        }
        return { trangthaithanhtoan, ...rest };
    });

    filtered_data.push({
        against_sales_order: null,
        customer_name: "Tổng tiền",
        name: null,
        "select_item": null,
        "total": Total,
        "trangthaithanhtoan": null
    })

    frappe.render_grid({
        title: "Báo cáo công nợ theo đơn hàng",
        subtitle: "Báo cáo công nợ theo đơn hàng",
        print_settings: print_settings,
        columns: columns,
        data: filtered_data,
        can_use_smaller_font: 1,
    });
}
